src/main/java/com/gojek/beast/sink/bq/BqSink.java

```diff
@@ -29,48 +29,93 @@
     private final BigQuery bigquery;
     private final TableId tableId;
     private final BQResponseParser responseParser;
-    private final BQErrorHandler errorHandler; // handler instance for BQ related errors
     private final BQRow recordInserter;
+    private final ErrorWriter errorWriter;
 
     private final Stats statsClient = Stats.client();
 
+    private final BQResponseFilter retryableInsertsFilter;
+    private final BQResponseFilter oobInsertsFilter;
+    private final BQResponseFilter unhandledInsertsFilter;
+
     @Override
     public Status push(Records records) {
-        InsertAllResponse response = insertIntoBQ(records);
-        InsertStatus status = new InsertStatus(!response.hasErrors(), response.getInsertErrors());
+        // filter invalid records which cannot be sunk to BQ
+        Map<BQRecordsErrorType, List<Record>> dlqSinkableRecords = new HashMap<>();
+        List<Record> validRecordList = new ArrayList<>();
+        records.forEach(record -> {
+            if (record.getColumns().isEmpty()) {
+                List<Record> errList = dlqSinkableRecords.getOrDefault(BQRecordsErrorType.DESERIALIZE, new ArrayList<>());
+                errList.add(record);
+                dlqSinkableRecords.put(BQRecordsErrorType.DESERIALIZE, errList);
+            } else {
+                validRecordList.add(record);
+            }
+        });
```

Comment:
Yes, I thought of this as well, actually not just processing time it will also increase memory requirements because each `Record` has key and value as raw bytes embedded into it now. What I was thinking earlier is to proactively check if the partition key value has OOB values(because we already know what range works for BQ), it's like adding sanity filters before pushing records for processing/insertion. As you said we can do this in `ConsumerRecordConverter`.

Or let's say we don't check for OOB records in `ConsumerRecordConverter` and let it go and fail first(only handle deserialize errors), side effects of that are GCS will have two files for a single batch(if it contains both null and OOB errors, which is very rare but it is possible) & we will have to keep message raw bytes and have increased memory footprint. What do you suggest? @chakravarthyvp

Reply:
My suggestion would be:
a) The `ConsumerRecordConverter` will push deserialization failures ONLY.
b) the notion of OOB is outside the converter ie., the OOB term is for BQSink and therefore we should store OOB records in DLQ in BQSink class as is today.
In fairness, the overall implementation becomes easy too. You just need the record writer wired into the ConsumerRecordConverter when you create in the Factory.

Reply:
As the name suggests I think `ConsumerRecordConverter` sole purpose should be conversion and Sinks should handle what to do with valid/invalid messages. One way to avoid that `for` loop and not change the responsibility of the converter as well as is to change the sink interface to pass valid/invalid records separately. Writing to DLQ from multiple places looks lesser cleaner to me.

Reply:
The `ErrorHandler#handleErrors()` enables abstraction and implementation to be injected anywhere in the code. So whether it is in the ConsumerRecordConverter or BQSink or OffsetCommit, the ErrorHandler hides the details of a DLQ. (It can log to console, log to a file, sink to GCS or new implementation can be provided to push to kafka). As far as this detail is not exposed to let's say to ConsumerRecordInverter or BQSink we can pass the handlers anywhere. That was the purpose of introducing that interface.

Think of it has a `Logger#error()` it does not matter where it's used (sink/converter etc.). How/Where it logs that's an implementation detail hidden (console/file) from the business logic classes.  Hope it helps. We can get @mauliksoneji @sravankorumilli's opinion as well.